faunadb/functions_basic.go

```diff
@@ -0,0 +1,291 @@
+
+type doFn struct {
+	fnApply
+	Do Expr `json:"do" faunarepr:"varargs"`
+}
+
+func (fn doFn) String() string { return printFn(fn) }
+
+// If evaluates and returns then or elze depending on the value of cond.
+// If cond evaluates to anything other than a boolean, if returns an “invalid argument” error
+//
+// Parameters:
+//  cond bool - A boolean expression.
+//  then Expr - The expression to run if condition is true.
+//  elze Expr - The expression to run if condition is false.
+//
+// Returns:
+//  Value - The result of either then or elze expression.
+//
+// See: https://app.fauna.com/documentation/reference/queryapi#basic-forms
+func If(cond, then, elze interface{}) Expr {
+	return ifFn{
+		If:   wrap(cond),
+		Then: wrap(then),
+		Else: wrap(elze),
+	}
+}
+
+type ifFn struct {
+	fnApply
+	If   Expr `json:"if"`
+	Then Expr `json:"then"`
+	Else Expr `json:"else"`
+}
+
+func (fn ifFn) String() string { return printFn(fn) }
+
+// Lambda creates an anonymous function. Mostly used with Collection functions.
+//
+// Parameters:
+//  varName string|[]string - A string or an array of strings of arguments name to be bound in the body of the lambda.
+//  expr Expr - An expression used as the body of the lambda.
+//
+// Returns:
+//  Value - The result of the body expression.
+//
+// See: https://app.fauna.com/documentation/reference/queryapi#basic-forms
+func Lambda(varName, expr interface{}) Expr {
+	return lambdaFn{
+		Lambda:     wrap(varName),
+		Expression: wrap(expr),
+	}
+}
+
+type lambdaFn struct {
+	fnApply
+	Lambda     Expr `json:"lambda"`
+	Expression Expr `json:"expr"`
+}
+
+func (fn lambdaFn) String() string { return printFn(fn) }
+
+// At execute an expression at a given timestamp.
+//
+// Parameters:
+//  timestamp time - The timestamp in which the expression will be evaluated.
+//  expr Expr - An expression to be evaluated.
+//
+// Returns:
+//  Value - The result of the given expression.
+//
+// See: https://app.fauna.com/documentation/reference/queryapi#basic-forms
+func At(timestamp, expr interface{}) Expr {
+	return atFn{
+		At:         wrap(timestamp),
+		Expression: wrap(expr),
+	}
+}
+
+type atFn struct {
+	fnApply
+	At         Expr `json:"at"`
+	Expression Expr `json:"expr"`
+}
+
+func (fn atFn) String() string { return printFn(fn) }
+
+// LetBuilder builds Let expressions
+type LetBuilder struct {
+	bindings unescapedArr
+}
+
+type letFn struct {
+	fnApply
+	Let Expr `json:"let"`
+	In  Expr `json:"in"`
+}
+
+func (fn letFn) String() string {
+	bindings := []string{}
+	tpe := reflect.TypeOf(fn.Let)
+	if tpe == reflect.TypeOf(unescapedArr{}) {
+		arr := fn.Let.(unescapedArr)
+		for _, elem := range arr {
+			m := elem.(unescapedObj)
+			for k, v := range m {
+				binding := "Bind(" + strconv.Quote(k) + ", " + v.String() + ")"
+				bindings = append(bindings, binding)
+			}
+		}
+	} else {
+		panic(fmt.Sprintf("Let bindings should be an unescapedArr, %s instead", tpe))
```

Comment:
Panicking in string functions is a bit unexpected. I'd recommend to put a "unexpected/invalid type" message rather than firing an unrecoverable error.